remove redundant GetDataMsg code in procMsg

palmskog opened the pull request:
This pull request both proposes a fix and raises a general issue (question).

Consider that the assignment `let: matchingBlocks := [:: get_block bt h] in` in `Protocol.v` always leaves `matchingBlocks` as a non-empty list. This means that pattern matching on `ohead matchingBlocks` will always take the `Some(b)` branch, no matter what. In turn, this implies that all the code related to `matchingTxs` is redundant.

As this pull request shows, all the theorems still hold with `matchingTxs` code excised. But is this really the intention, i.e., does it capture real-world blockchain protocol behavior? I'd be interested in any ideas or pointers there.

dranov commented:
Hi @palmskog. Thanks for spotting this.

The problem here is that `matchingBlocks`, i.e. `let: matchingBlocks := [:: get_block bt h]` has incorrect behaviour. Up until [this commit](https://github.com/certichain/toychain/commit/f211dd5b825dc83ce2d97feda1748650763d20fa#diff-ec331168342c773f76226823f9237545R222) (which contains fixes needed after we replaced our blocktree axioms with an actual implementation), this was `let: matchingBlocks := [seq b <- bt | (hashB b) == h]` – what it was intended to be.

The intuition there was that we assumed hashes for blocks would not collide with hashes for transactions, and this meant that at most one of `matchingBlocks := [seq b <- bt | (hashB b) == h]` or `matchingTxs := [seq t <- pool | (hashT t) == h]` would be non-empty.

The reason all the proofs still work even if the behaviour of `GetDataMsg` is completely borked (as it is) is that we don't use this part of the protocol for any of the current proofs. Our current invariant works in a clique topology – blocks and transactions are directly broadcast to all participants, so we don't need to reason about `Inv` and `GetData`.

I don't recall exactly why I made the change, though. I'll take a look tomorrow and report back.

dranov commented:
`let: matchingBlocks := [seq b <- [:: get_block bt h] | b != GenesisBlock] in` looks like a reasonable fix. Some proofs do break (including in the invariant where it mentions `GetDataMsg`), but not in a major way.

palmskog commented:
@dranov I updated my PR with your suggested fix for `matchingBlocks`, and updated the proofs. Can you please confirm that this is what you had in mind?

Note that you can use "handler case lemma+tactic" approach from Verdi to separate protocol code from invariant proofs. For example, here is the [handler case lemma](https://github.com/DistributedComponents/verdi-lockserv/blob/master/systems/LockServ.v#L240) for a simple system, and here is the [corresponding tactic](https://github.com/DistributedComponents/verdi-lockserv/blob/master/systems/LockServ.v#L277).

dranov commented:
Looks good! I'll take a closer look at the handler case lemma+tactic approach. Thanks.
